File: android/app/src/main/java/com/herobrawl/game/engine/Idle.kt
package com.herobrawl.game.engine

import com.herobrawl.game.model.GameState
import com.herobrawl.game.model.IdleRates
import kotlin.math.roundToLong

data class IdleReward(
    val gold: Long,
    val spirit: Long,
    val shards: Long,
    val hours: Double,
    val capped: Boolean,
)

object Idle {
    fun compute(state: GameState, now: Long): IdleReward {
        val elapsedMs = (now - state.idle.tickedAt).coerceAtLeast(0L)
        val capMs = state.idle.capHours * 3600L * 1000L
        val effective = minOf(elapsedMs, capMs)
        val hours = effective / 3_600_000.0
        return IdleReward(
            gold = (state.idle.ratePerHour.gold * hours).roundToLong(),
            spirit = (state.idle.ratePerHour.spirit * hours).roundToLong(),
            shards = (state.idle.ratePerHour.shards * hours).roundToLong(),
            hours = hours,
            capped = elapsedMs >= capMs,
        )
    }

    fun claim(state: GameState, now: Long): GameState {
        val r = compute(state, now)
        return state.copy(
            idle = state.idle.copy(tickedAt = now),
            currency = state.currency.copy(
                gold = state.currency.gold + r.gold,
                spirit = state.currency.spirit + r.spirit,
                shards = state.currency.shards + r.shards,
            )
        )
    }

    fun upgradedRates(chapter: Int, stage: Int): IdleRates {
        val progress = (chapter - 1) * 10 + (stage - 1)
        val mul = 1.0 + progress * 0.08
        return IdleRates(
            gold = (1200L * mul).roundToLong(),
            spirit = (60L * mul).roundToLong(),
            shards = (4L * mul).roundToLong(),
        )
    }
}

File: src/game/idle.ts
import type { GameState, IdleRates } from "../types";

export interface IdleReward {
  gold: number;
  spirit: number;
  shards: number;
  hours: number;
  capped: boolean;
}

export function computeIdleReward(state: GameState, now = Date.now()): IdleReward {
  const elapsedMs = now - state.idle.tickedAt;
  const capMs = state.idle.capHours * 3600 * 1000;
  const effective = Math.max(0, Math.min(elapsedMs, capMs));
  const hours = effective / 3600000;

  const { gold, spirit, shards } = state.idle.ratePerHour;

  return {
    gold: Math.floor(gold * hours),
    spirit: Math.floor(spirit * hours),
    shards: Math.floor(shards * hours),
    hours,
    capped: elapsedMs >= capMs && capMs > 0,
  };
}

export function claimIdleReward(state: GameState, now = Date.now()): GameState {
  const r = computeIdleReward(state, now);
  return {
    ...state,
    idle: { ...state.idle, tickedAt: now },
    currency: {
      ...state.currency,
      gold: state.currency.gold + r.gold,
      spirit: state.currency.spirit + r.spirit,
      shards: state.currency.shards + r.shards,
    },
  };
}

export function upgradeIdleRates(base: IdleRates, chapter: number, stage: number): IdleRates {
  const progress = (chapter - 1) * 10 + (stage - 1);
  const mul = 1 + progress * 0.08;
  return {
    gold: Math.round(base.gold * mul),
    spirit: Math.round(base.spirit * mul),
    shards: Math.round(base.shards * mul),
  };
}
